### ISSUE TITLE
hh_server hangs forever on large switch statment

### ISSUE THREAD

**Issue by ssandler** [facebook/hhvm#8180]
### HHVM Version
3.24.6 

### Operating System and Version
Reproduced on OSX and Ubuntu 16.04

### Standalone code, or other way to reproduce the problem
This code reproduces it even if it's the only file in the repo, see https://github.com/ssandler/hhvm_switch_bug for a repro.

The code does do the right thing at runtime in HHVM, and `hh_parse` doesn't seem to complain. It's just hh_server that hangs at the "naming" step. I let it run for a few hours to see if it would ever finish.

### Expected result
`hh_server` succeeds

### Actual result
`hh_server` hangs forever

**Comment by fredemmott** [comment]
Repros on nightlies too.

**Comment by fredemmott** [comment]
Internal task: t28831714

**Comment by lexidor** [comment]
I am going over old issues on this repository, to see which ones apply to the current versions of hhvm.

I am happy to report that this is fixed (tested in hhvm 4.56).

```
$ git clone https://github.com/ssandler/hhvm_switch_bug
Cloning into 'hhvm_switch_bug'...
remote: Enumerating objects: 10, done.
remote: Total 10 (delta 0), reused 0 (delta 0), pack-reused 10
Unpacking objects: 100% (10/10), 11.11 KiB | 2.78 MiB/s, done.

$ cd hhvm_switch_bug/

$ time hh_client 
For more detailed logs, try `tail -f $(hh_client --monitor-logname) $(hh_client --logname)`
Server launched with the following command:
        '/opt/hhvm/4.56.0/bin/hh_server' '-d' '/path/to/hhvm_switch_bug' '--waiting-client' '8'
Spawned typechecker (child pid=14308)
Logs will go to /tmp/hh_server/monitor_logs/...hhvm_switch_bug-2020-06-02-16-04-31.monitor_log
No errors!

real    0m0,845s
user    0m0,086s
sys     0m0,017s
```
